engine/table/src/main/java/io/deephaven/engine/table/impl/PartitionAwareSourceTable.java

```diff
@@ -263,30 +290,54 @@ protected final Collection<LiveSupplier<ImmutableTableLocationKey>> filterLocati
 
     @Override
     public Table where(Filter filter) {
-        return whereImpl(WhereFilter.fromInternal(filter));
+        return whereImpl(ExtractInnerConjunctiveFilters.of(ConjunctiveFilter.of(WhereFilter.fromInternal(filter))));
     }
 
-    private Table whereImpl(final WhereFilter[] whereFilters) {
-        if (whereFilters.length == 0) {
+    private Table whereImpl(final List<WhereFilter> whereFilters) {
+        if (whereFilters.isEmpty()) {
             return prepareReturnThis();
         }
 
         final QueryCompilerRequestProcessor.BatchProcessor compilationProcessor = QueryCompilerRequestProcessor.batch();
         final List<WhereFilter> partitionFilters = new ArrayList<>();
         final List<WhereFilter> otherFilters = new ArrayList<>();
+
+        boolean serialFilterFound = false;
+        boolean partitioningFilterFound = false;
+        final Set<Object> partitionBarriers = new HashSet<>();
         for (WhereFilter whereFilter : whereFilters) {
             whereFilter.init(definition, compilationProcessor);
-            if (!(whereFilter instanceof ReindexingFilter)
-                    && isValidAgainstColumnPartitionTable(whereFilter.getColumns(), whereFilter.getColumnArrays())) {
+
+            if (!whereFilter.permitParallelization()) {
+                serialFilterFound = true;
+            }
+
+            final boolean isPartitioningFilter = !(whereFilter instanceof ReindexingFilter)
+                    && isValidAgainstColumnPartitionTable(whereFilter.getColumns(), whereFilter.getColumnArrays());
+            partitioningFilterFound |= isPartitioningFilter;
+
+            final boolean missingBarrier = !partitionBarriers.containsAll(ExtractRespectedBarriers.of(whereFilter));
+            if (serialFilterFound || missingBarrier) {
```

Comment:
I had to puzzle this out for myself, so I suggest adding some comments explaining what we are doing.  The could go separate closer to the check (e.g., where we set serialFilterFound); or reword them to your liking.

```suggestion
            // once we've found a serial filter, then we cannot prioritize any filter and we put every filter (including the partitioning filters) into the post coalescing filter set.
            // similarly, anytime we prioritize a partitioning filter, we record the barriers that it declares.  A filter that respects no barriers, or only those prioritized barriers may also be prioritized.  A filter that respects any barrier which was not in partition filters (meaning it must be in in otherFilters - because otherwise you would be respecting an undeclared barrier); cannot be prioritized because that would jump the barrier.
            if (serialFilterFound || missingBarrier) {
```